### ISSUE TITLE
Add an option to use simplified collider mesh to Bake Meshes ProtoFlux node

### ISSUE THREAD

**Issue by JackTheFoxOtter** [Yellow-Dog-Man/Resonite-Issues#1801]
### Is your feature request related to a problem? Please describe.

Many creations in Resonite have use a simplified collision shape, separate to the rendered mesh. Often those are using primitive colliders instead of mesh colliders. Currently, when using the `BakeMeshes` ProtoFlux node, if you configure it to create a collider for the baked result, it will always pick a mesh collider using the same baked mesh that's also used in the MeshRenderer.  

### Describe the solution you'd like

I would like an option to use simplified collision shapes (if available) when baking a mesh. When using that option the node would bake two meshes, the visual mesh for the `MeshRenderer`, and a collision mesh for the `MeshCollider`. The collision mesh would be composed of all primitive colliders and meshes of `MeshColliders`, while the visual mesh would be composed of the meshes used in `MeshRenderer` components (the current behavior).

### Describe alternatives you've considered

Manually creating a simplified collision mesh after the bake, or just using the non-simplified collider it already generates.

**Comment by mpmxyz** [comment]
> I'm kind of confused by this issue- isn't the solution to set ColliderHandling to NeverAdd if the user has already configured their colliders how they need them?
I think that the issue arises when you want to bake arbitrary objects and not just objects you have full control over.

Imagine you want to bake a scene with a table and a few people sitting on chairs drinking from mugs.
Before baking the colliders would be quite optimized:
- The table may use a `BoxCollider` or `CylinderCollider`.
- The avatars use `CapsuleColliders` and a few auxilliary ones.
- The chairs are simple `BoxColliders`.
- The mugs are a single `CyliderCollider` each.

When you bake this scene you are suddenly confronted with a MeshCollder using a high-poly mesh.
Interacting with it is going to freeze you when the collider is primed and drop framerate while you are nearby.

**Comment by shiftyscales** [comment]
@JackTheFoxOtter - For your specific use-case, is it required that the creation of simplified colliders be done in realtime a part of your request? Or is it okay to just have better processes for handling it after the fact / outside of the function of the BakeMeshes node?

On mesh colliders there is already a "convex hull decomposition" option to split mesh colliders into simplified convex hulls that approximate the same shape. Is that what you had meant by "manually creating a simplified collision shape"?

@mpmxyz - As noted - it is already possible to run a convex hull decomposition on a mesh collider to split it into simpler colliders as-needed.

Convex hull decomposition can requires its values to be adjusted by the user to get the best results from it depending on what the user is after- e.g. the desired precision/closeness to the original shape.

I'm not sure how viable / helpful it would be to include this functionality into the BakeMeshes node itself since to get the best results from it- one would typically want to manually optimize/set up the simplified colliders anyway.

For further clarification so I better understand this issue, what do you mean when you say "simplified collision shapes", @JackTheFoxOtter? Are you talking about convex hulls? Or do you mean other collider primitives e.g. box, capsule, cone, etc.

**Comment by JackTheFoxOtter** [comment]
> I'm kind of confused by this issue- isn't the solution to set ColliderHandling to NeverAdd if the user has already configured their colliders how they need them?

No. If you don't add a collider to the baked mesh, it won't have a collider, and you can't add one after the fact without inspecting the object you've created. There's no way to do this with ProtoFlux currently, and if there was, I'd still request a simpler way of doing it.

> @JackTheFoxOtter - For your specific use-case, is it required that the creation of simplified colliders be done in realtime a part of your request? Or is it okay to just have better processes for handling it after the fact / outside of the function of the BakeMeshes node?

I'm not sure what you mean by "real time". My request was for the ability to have the `BakeMeshes` node to create two separate mesh, one mesh for all visual meshes baked together, and one for all (primitive or mesh) collider meshes baked together. The end result would still be a mesh with a mesh collider, the difference would be that the collider mesh and the visual mesh would be different.

> @mpmxyz - As noted - it is already possible to run a convex hull decomposition on a mesh collider to split it into simpler colliders as-needed.

Yeah, but that's not what my request was about.

> I'm not sure how viable / helpful it would be to include this functionality into the BakeMeshes node itself since to get the best results from it- one would typically want to manually optimize/set up the simplified colliders anyway.

That would be a separate issue, and much more useful as an entirely separate node. But your assumption "one would typically want to manully..." doesn't apply here, as this wouldn't be an automated solution, which is what I'm requesting.

> For further clarification so I better understand this issue, what do you mean when you say "simplified collision shapes", @JackTheFoxOtter? Are you talking about convex hulls? Or do you mean other collider primitives e.g. box, capsule, cone, etc.

A separate collider mesh baked from all existing colliders on the target hierarchy. That would include all meshes used in `MeshCollider` components as well as mesh representations of all primitive collider components. My point here was that many optimized creations don't use the same collision shape as the visual one. I'd like an option to account for that with the `BakeMeshes` node.

My specific use-case was the LiteBlock tool. It allows you to bake your creations into a static mesh. The problem there is that right now, I can only either add no collider at all, or use the baked result mesh as the collider.

Current behavior of the tool:
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/35171546/1a46331c-24c0-42da-a997-a07568a028f3)
The `MeshRenderer` and `MeshCollider` both reference the same, baked `StaticMesh`.

With collider handling set to don't add:
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/35171546/5e0631e9-f6ee-41db-b67c-bd74b9ce7727)
No collider is added, the object can't be picked up.

Now, to demonstrate what I would like to automate, I've manually replaced all `BoxColliders` with `BoxMeshes` and assignd those to the `MeshRenderer`s, then baked with collider handling set to always add:
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/35171546/85ac98a9-a691-4f3f-be7c-e80a0c8600ea)
The result is a low-poly mesh, reflecting the original collision shape of the individual pieces. Obviously, this doesn't look very nice as the rendered mesh, but it represents the original collission shape as a mesh, which I'd like to use as the collission mesh alongside the baked visual mesh for the final baked result.

**Comment by shiftyscales** [comment]
I think I better understand your request now- thanks for the additional clarification, @JackTheFoxOtter.

I think your specific request would probably be better suited to the introduction of a "Bake Colliders" node instead, @JackTheFoxOtter.

Seeking input from @Frooxius or @ProbablePrime.

**Comment by ProbablePrime** [comment]
I defer to @Frooxius I'm not sure if we'd want to do this.

**Comment by Frooxius** [comment]
If I understand it correctly, you essentially want the baking to:
- Go through visual meshes and bake those together
- Go through collider meshes in the hierarchy and bake those meshes together

It sounds like you might want primitive colliders too, but we can't really bake those into mesh colliders - they work quite differently. 

What we could do though is that it bakes all the mesh colliders separately and then just copies all the primitive colliders as is (or makes a compound collider once that's supported).

**Comment by JackTheFoxOtter** [comment]
Yes, that would work! And be even better because you'd maintain the primitive colliders as-is. If the object would only contain primitive colliders they'd be copied to the baked result, and no mesh collider would be added.
